File: 1_Sorting/3_insertion_Sort/insertionSort.cpp
// #include<iostream>
// using namespace std;

// int main(){
//     int arr[1000];
//     int n;

//     cout << "Enter the size of array: ";
//     cin >> n;

//     for (int i = 0; i < n; i++){
//         cin >> arr[i];
//     }

//     for (int i = 1; i < n; i++){

//         for(int j = i; j > 0; j--){

//             if (arr[j] < arr[j-1]){
//                 swap(arr[j], arr[j-1]);
//             }
//             else{
//                 break;
//             }
//         }
//     }

//     for (int i = 0; i < n; i++){
//         cout << arr[i] << " ";
//     }
//     cout << endl;

//     return 0;
// }




// ---------------- --------------

#include <iostream>
#include<vector>
using namespace std;

void insertionSort(vector<int>& arr){
    int n = arr.size();

    for(int i = 1; i < n; ++i){
        int key = arr[i];
        int j = i - 1;

        while(j >= 0 && arr[j] > key){
            arr[j + 1] = arr[j];
            j--;
        }

        arr[j + 1] = key;
    }
}

void printArray(const vector<int>& arr){
    for(int num : arr){
        cout << num << " ";
    }
    cout << "\n";
}

int main(){
    vector<int> arr = {8, 5, 1, 4, 3};

    cout << "Original array:\n";
    printArray(arr);

    insertionSort(arr);

    cout << "Sorted array:\n";
    printArray(arr);

    return 0;
}

File: 1_Sorting/3_insertion_Sort/insertionSort.js
function insertionSort(arr){
    for (let i = 1; i < arr.length; i++){
        let key = arr[i];
        let j = i - 1;

        // Move elements that are greater than key to one position ahead
        while (j >= 0 && arr[j] > key){
            arr[j + 1] = arr[j];
            j--;
        }

        // Insert the key into ints correct positiono
        arr[j + 1] = key;
    }

    return arr;
}

const array = [5, 2, 9, 1, 5, 6];
console.log("Before sorting: ", array);

const sorted = insertionSort(array);
console.log("After sorting: ", sorted);

// @Dheeraj2002kumar ➜ .../DSA-WORDSPACE/DSA /1_Sorting /3_insertion_Sort(main) $ node insertionSort.js
// Before sorting: [5, 2, 9, 1, 5, 6]
// After sorting: [1, 2, 5, 5, 6, 9]
// @Dheeraj2002kumar ➜ .../DSA-WORDSPACE/DSA /1_Sorting /3_insertion_Sort(main) $ 